### PR TITLE
TLoZ: Implementing The Legend of Zelda

### PR DESCRIPTION
**Pull request by Rosalie-A**
Please format your title with what portion of the project this pull request is
targeting and what it's changing.

ex. "MyGame4: implement new game" or "Docs: add new guide for customizing MyGame3"

## What is this fixing or adding?
Support for _The Legend of Zelda_ for the NES, including randomizer, client, and emulator connector, as well as website documentation.

N.B.: while I'm the one putting in the pull request, @t3hf1gm3nt put in a nontrivial amount of work and as such deserves joint credit.

## How was this tested?
Public and private sync beta tests, with multiple players playing both Z1 and non-Z1 games.

### CODE REVIEW

File: host.yaml [hunk 1]
```diff
@@ -125,6 +125,18 @@ soe_options:
...
+tloz_options:
+  # File name of the Zelda 1
+  rom_file: "Legend of Zelda, The (U) (PRG0) [!].nes"
+  # Set this to false to never autostart a rom (such as after patching)
+  #             true  for operating system default program
+  # Alternatively, a path to a program to open the .nes file with
+  rom_start: true
+  # Display message inside of Bizhawk
+  display_msgs: true
+smz3_options:
```
[the remark this answers is not in the document]
Comment (t3hf1gm3nt) [reply host.yaml:137]: I don't know why but the smz3 options kept disappearing from the host yaml, and them not being there kept causing the unit tests to yell.

[the remark this answers is not in the document]
Comment (Rosalie-A) [reply host.yaml]: In any event, I've removed the stray option and the unit tests seem to be fine, and I've also implemented a level 9 junk fill based on feedback from the latest beta test.

File: Zelda1Client.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+item_ids = item_game_ids
+location_ids = location_ids
+items_by_id = {id: item for item, id in item_ids.items()}
+locations_by_id = {id: location for location, id in location_ids.items()}
+
+
+class ZeldaCommandProcessor(ClientCommandProcessor):
+    def __init__(self, ctx: CommonContext):
+        super().__init__(ctx)
+
```
Comment (Berserker66) [inline Zelda1Client.py:44]:
```suggestion
```
These lines do nothing and can be removed

File: Zelda1Client.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        super().__init__(server_address, password)
+        self.bonus_items = []
+        self.nes_streams: (StreamReader, StreamWriter) = None
+        self.nes_sync_task = None
+        self.messages = {}
+        self.locations_array = None
+        self.nes_status = CONNECTION_INITIAL_STATUS
+        self.game = 'The Legend of Zelda'
+        self.awaiting_rom = False
+        self.display_msgs = True
```
Comment (Berserker66) [inline Zelda1Client.py:74]: This is a duplicate of a global, though I'd prefer the attribute is the one that stays.
I also don't see where either is polled, only set.

File: Zelda1Client.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        await self.send_connect()
+
+    def _set_message(self, msg: str, msg_id: int):
+        if DISPLAY_MSGS:
+            self.messages[(time.time(), msg_id)] = msg
+
+    def on_package(self, cmd: str, args: dict):
+        if cmd == 'Connected':
+            if "slot_data" in args.keys():
+                self.slot_data = args["slot_data"]
```
Comment (Berserker66) [inline Zelda1Client.py:98]:
```suggestion
            self.slot_data = args.get("slot_data", {})
```

File: Zelda1Client.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+                            short_name = "TakeAnyMiddle"
+                        if "Right" in location_name:
+                            short_name = "TakeAnyRight"
+                        if short_name is not None:
+                            item_code = ctx.slot_data[short_name]
+                            if item_code > 0:
+                                ctx.bonus_items.append(item_code)
+                            locations_checked.append(location)
+        if locations_checked:
+            print([ctx.location_names[location] for location in locations_checked])
```
Comment (Berserker66) [inline Zelda1Client.py:250]: This feels like a debug print rather than something for the user.

File: Zelda1Client.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+                    ctx.shop_slots[context_slot] |= shop_bit
+                if locations_array["takeAnys"] and locations_array["takeAnys"] >= 4:
+                    if "Take Any" in location_name:
+                        short_name = None
+                        if "Left" in location_name:
+                            short_name = "TakeAnyLeft"
+                        if "Middle" in location_name:
+                            short_name = "TakeAnyMiddle"
+                        if "Right" in location_name:
+                            short_name = "TakeAnyRight"
```
Comment (Berserker66) [inline Zelda1Client.py:243]: I'm assuming that an item can't be both Left and Middle, please use elif
```suggestion
                        if "Left" in location_name:
                            short_name = "TakeAnyLeft"
                        elif "Middle" in location_name:
                            short_name = "TakeAnyMiddle"
                        elif "Right" in location_name:
                            short_name = "TakeAnyRight"
```

File: Zelda1Client.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        if auto_start is True:
+            import webbrowser
+            webbrowser.open(romfile)
+        elif isinstance(auto_start, str) and os.path.isfile(auto_start):
+            subprocess.Popen([auto_start, romfile],
+                             stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+
+
+    async def main(args):
+        print(args)
```
Comment (Berserker66) [inline Zelda1Client.py:365]:
```suggestion
```

debug print should probably go. Or use logging.debug

File: data/lua/TLoZ/json.lua [hunk 8]
```diff
@@ -0,0 +1,380 @@
+--
+-- json.lua
+--
+-- Copyright (c) 2015 rxi
+--
+-- This library is free software; you can redistribute it and/or modify it
+-- under the terms of the MIT license. See LICENSE for details.
```
Comment (Berserker66) [inline data/lua/TLoZ/json.lua:7]: the MIT License doesn't have many requirements, but the one it does have is for the license file to accompany this file, which it does not in this PR.

File: worlds/tloz/Rules.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
+import typing
```
Comment (Berserker66) [inline worlds/tloz/Rules.py:1]: this entire file appears to be a copy-paste from the generic world, please just import that instead.

File: worlds/tloz/__init__.py [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+            # We shuffle the tiers of rupee caves. Caves that shared a value before still will.
+            secret_caves = self.multiworld.random.sample(sorted(secret_money_ids), 3)
+            secret_cave_money_amounts = [20, 50, 100]
+            for i, amount in enumerate(secret_cave_money_amounts):
+                # Giving approximately double the money to keep grinding down
+                amount = amount * self.multiworld.random.triangular(1.5, 2.5)
+                secret_cave_money_amounts[i] = int(amount)
+            for i, cave in enumerate(secret_caves):
+                rom_data[secret_money_ids[cave]] = secret_cave_money_amounts[i]
+            return rom_data
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:391]:
```suggestion
        with open(get_base_rom_path(), 'rb') as rom:
            rom_data = self.apply_base_patch(rom)
            # Write each location's new data in
        for location in self.multiworld.get_filled_locations(self.player):
            # Zelda and Ganon aren't real locations
            if location.name == "Ganon" or location.name == "Zelda":
                continue
        
            # Neither are boss defeat events
            if "Status" in location.name:
                continue
        
            item = location.item.name
            # Remote items are always going to look like Rupees.
            if location.item.player != self.player:
                item = "Rupee"
        
            item_id = item_game_ids[item]
            location_id = location_ids[location.name]
        
            # Shop prices need to be set
            if location.name in shop_locations:
                if location.name[-5:] == "Right":
                    # Final item in stores has bit 6 and 7 set. It's what marks the cave a shop.
                    item_id = item_id | 0b11000000
                price_location = shop_price_location_ids[location.name]
                item_price = item_prices[item]
                if item == "Rupee":
                    item_class = location.item.classification
                    if item_class == ItemClassification.progression:
                        item_price = item_price * 2
                    elif item_class == ItemClassification.useful:
                        item_price = item_price // 2
                    elif item_class == ItemClassification.filler:
                        item_price = item_price // 2
                    elif item_class == ItemClassification.trap:
                        item_price = item_price * 2
                rom_data[price_location] = item_price
            if location.name == "Take Any Item Right":
                # Same story as above: bit 6 is what makes this a Take Any cave
                item_id = item_id | 0b01000000
            rom_data[location_id] = item_id
        
        # We shuffle the tiers of rupee caves. Caves that shared a value before still will.
        secret_caves = self.multiworld.random.sample(sorted(secret_money_ids), 3)
        secret_cave_money_amounts = [20, 50, 100]
        for i, amount in enumerate(secret_cave_money_amounts):
            # Giving approximately double the money to keep grinding down
            amount = amount * self.multiworld.random.triangular(1.5, 2.5)
            secret_cave_money_amounts[i] = int(amount)
        for i, cave in enumerate(secret_caves):
            rom_data[secret_money_ids[cave]] = secret_cave_money_amounts[i]
        return rom_data
```
Closes the file earlier and moves the player filtering into the function call.

File: worlds/tloz/__init__.py [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+
+        patched_filename = os.path.join(output_directory, outputFilename)
+
+        with open(patched_filename, 'wb') as patched_rom_file:
+            patched_rom_file.write(patched_rom)
+
+        patch = TLoZDeltaPatch(os.path.splitext(outputFilename)[0] + TLoZDeltaPatch.patch_file_ending,
+                               player=self.player,
+                               player_name=self.multiworld.player_name[self.player], patched_path=outputFilename)
+        self.rom_name_available_event.set()
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:418]:
```suggestion
        try:
            patched_rom = self.apply_randomizer()
            outfilebase = 'AP_' + self.multiworld.seed_name
            outfilepname = f'_P{self.player}'
            outfilepname += f"_{self.multiworld.get_file_safe_player_name(self.player).replace(' ', '_')}"
            outputFilename = os.path.join(output_directory, f'{outfilebase}{outfilepname}.nes')

            self.rom_name_text = f'LOZ{Utils.__version__.replace(".", "")[0:3]}_{self.player}_{self.multiworld.seed:11}\0'
            self.romName = bytearray(self.rom_name_text, 'utf8')[:0x20]
            self.romName.extend([0] * (0x20 - len(self.romName)))
            self.rom_name = self.romName
            patched_rom[0x10:0x30] = self.romName

            self.playerName = bytearray(self.multiworld.player_name[self.player], 'utf8')[:0x20]
            self.playerName.extend([0] * (0x20 - len(self.playerName)))
            patched_rom[0x30:0x50] = self.playerName

            patched_filename = os.path.join(output_directory, outputFilename)

            with open(patched_filename, 'wb') as patched_rom_file:
                patched_rom_file.write(patched_rom)

            patch = TLoZDeltaPatch(os.path.splitext(outputFilename)[0] + TLoZDeltaPatch.patch_file_ending,
                                   player=self.player,
                                   player_name=self.multiworld.player_name[self.player], patched_path=outputFilename)
        finally:
            self.rom_name_available_event.set()
```
ensures the event is always set and doesn't make the thread stuck

File: worlds/tloz/__init__.py [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+        patch.write()
+        os.unlink(outputFilename)
+
+    def modify_multidata(self, multidata: dict):
+        import base64
+        rom_name = getattr(self, "rom_name", None)
+        if rom_name:
+            new_name = base64.b64encode(bytes(self.rom_name)).decode()
+            multidata["connect_names"][new_name] = multidata["connect_names"][self.multiworld.player_name[self.player]]
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:427]:
```suggestion
        self.rom_name_available_event.wait()
        new_name = base64.b64encode(bytes(self.rom_name)).decode()
        multidata["connect_names"][new_name] = multidata["connect_names"][self.multiworld.player_name[self.player]]
```

Wait until the other thread has the name, then it it will also exist.

File: worlds/tloz/__init__.py [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+                        elif item_class == ItemClassification.trap:
+                            item_price = item_price * 2
+                    rom_data[price_location] = item_price
+                if location.name == "Take Any Item Right":
+                    # Same story as above: bit 6 is what makes this a Take Any cave
+                    item_id = item_id | 0b01000000
+                rom_data[location_id] = item_id
+
+            # We shuffle the tiers of rupee caves. Caves that shared a value before still will.
+            secret_caves = self.multiworld.random.sample(sorted(secret_money_ids), 3)
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:383]: apply_randomizer() is called from generate_output() which is threaded, accessing the global random gives non-deterministic results. Either randomize this ahead of time or use multiworld.slot_seeds[self.player] to have your own random instance.

File: worlds/tloz/__init__.py [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+
+    def modify_multidata(self, multidata: dict):
+        import base64
+        rom_name = getattr(self, "rom_name", None)
+        if rom_name:
+            new_name = base64.b64encode(bytes(self.rom_name)).decode()
+            multidata["connect_names"][new_name] = multidata["connect_names"][self.multiworld.player_name[self.player]]
+
+    def get_filler_item_name(self) -> str:
+        filler_items = [item for item in item_table if item_table[item].classification == ItemClassification.filler]
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:430]: this should be assembled once ahead of time, or on first need. This function can get called many many times.

File: worlds/tloz/__init__.py [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+
+class TLoZItem(Item):
+    game = 'The Legend of Zelda'
+
+
+class TLoZLocation(Location):
+    game = 'The Legend of Zelda'
+
+
+class PlandoItem(NamedTuple):
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:473]: why is this here?

File: worlds/tloz/__init__.py [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+            logging.debug(f'{warning}')
+
+    def failed(self, warning: str, exception=Exception):
+        if self.force in ['true', 'fail', 'failure']:
+            raise exception(warning)
+        else:
+            self.warn(warning)
+
+
+class PlandoConnection(NamedTuple):
```
Comment (Berserker66) [inline worlds/tloz/__init__.py:493]: Same as above

File: worlds/tloz/docs/en_The Legend of Zelda.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
+# The Legend of Zelda (NES)
+
+## Where is the settings page?
+
+The [player settings page for this game](../player-settings) contains all the options you need to configure and export a
+config file.
+
+## What does randomization do to this game?
+
+All acquirable pickups (except maps and compasses) are shuffled amongst each other. Logic is in place to ensure both
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/en_The Legend of Zelda.md:10]:
```suggestion
All acquirable pickups (except maps and compasses) are shuffled among each other. Logic is in place to ensure both
```
This isn't necessary by any means, but I prefer among to amongst when trying to write accessible material.

File: worlds/tloz/docs/en_The Legend of Zelda.md [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+that the game is still completable, and ensuring that players aren't forced to enter dungeons under-geared.
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/en_The Legend of Zelda.md:11]:
```suggestion
that the game is still completable, and that players aren't forced to enter dungeons under-geared.
```

File: worlds/tloz/docs/en_The Legend of Zelda.md [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+Shops can contain any item in the game, with prices added for the items unavailable in stores. Rupee caves are worth
+more while shops will cost less, making shop routing and money management important without requiring mindless grinding.
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/en_The Legend of Zelda.md:14]:
```suggestion
more while shops cost less, making shop routing and money management important without requiring mindless grinding.
```

File: worlds/tloz/docs/multiworld_en.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+- The Zelda1Client
+    - Bundled with Archipelago: [Archipelago Releases Page](https://github.com/ArchipelagoMW/Archipelago/releases)
+- The BizHawk emulator. Versions 2.3.1 and higher are supported. Version 2.7 is recommended
+    - [BizHawk Official Website](http://tasvideos.org/BizHawk.html)
+
+## Installation Procedures
+
+1. Download and install the latest version of Archipelago.
+    1. On Windows, download Setup.Archipelago.<HighestVersion\>.exe and run it
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:13]:
```suggestion
    - On Windows, download Setup.Archipelago.<HighestVersion\>.exe and run it.
```

File: worlds/tloz/docs/multiworld_en.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+2. Assign Bizhawk version 2.3.1 or higher as your default program for launching `.nes` files.
+    1. Extract your Bizhawk folder to your Desktop, or somewhere you will remember. Below are optional additional steps
+       for loading ROMs more conveniently
+        1. Right-click on a ROM file and select **Open with...**
+        2. Check the box next to **Always use this app to open .nes files**
+        3. Scroll to the bottom of the list and click the grey text **Look for another App on this PC**
+        4. Browse for `EmuHawk.exe` located inside your Bizhawk folder (from step 1) and click **Open**.
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:20]:
```suggestion
    - Extract your Bizhawk folder to your Desktop, or somewhere you will remember. Below are optional additional steps
       for loading ROMs more conveniently.
        1. Right-click on a ROM file and select **Open with...**
        2. Check the box next to **Always use this app to open .nes files**.
        3. Scroll to the bottom of the list and click the grey text **Look for another App on this PC**.
        4. Browse for `EmuHawk.exe` located inside your Bizhawk folder (from step 1) and click **Open**.
```

File: worlds/tloz/docs/multiworld_en.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+## Running the Client Program and Connecting to the Server
+
+Once the Archipelago server has been hosted:
+
+1. Navigate to your Archipelago install folder and run `ArchipelagoZelda1Client.exe`
+2. Notice the `/connect command` on the server hosting page (It should look like `/connect archipelago.gg:*****`
+   where ***** are numbers)
+3. Type the connect command into the client OR add the port to the pre-populated address on the top bar (it should
+   already say `archipelago.gg`) and click `connect`
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:70]:
```suggestion
1. Navigate to your Archipelago install folder and run `ArchipelagoZelda1Client.exe`.
2. Notice the `/connect command` on the server hosting page. (It should look like `/connect archipelago.gg:*****`
   where ***** are numbers)
3. Type the connect command into the client OR add the port to the pre-populated address on the top bar (it should
   already say `archipelago.gg`) and click `connect`.
```

File: worlds/tloz/docs/multiworld_en.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+
+1. Open Bizhawk 2.3.1 or higher and load your ROM OR click your ROM file if it is already associated with the
+   extension `*.nes`
+2. Click on the Tools menu and click on **Lua Console**
+3. Click the folder button to open a new Lua script. (CTL-O or **Script** -> **Open Script**)
+4. Navigate to the location you installed Archipelago to. Open data/lua/TLOZ/tloz_connector.lua
+    1. If it gives a `NLua.Exceptions.LuaScriptException: .\socket.lua:13: module 'socket.core' not found:` exception
+       close your emulator entirely, restart it and re-run these steps
+    2. If it says `Must use a version of bizhawk 2.3.1 or higher`, double-check your Bizhawk version by clicking **
+       Help** -> **About**
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:82]:
```suggestion
1. Open Bizhawk 2.3.1 or higher and load your ROM OR click your ROM file if it is already associated with the
   extension `*.nes`.
2. Click on the Tools menu and click on **Lua Console**.
3. Click the folder button to open a new Lua script. (CTL-O or **Script** -> **Open Script**)
4. Navigate to the location you installed Archipelago to. Open `data/lua/TLOZ/tloz_connector.lua`.
    1. If it gives a `NLua.Exceptions.LuaScriptException: .\socket.lua:13: module 'socket.core' not found:` exception
       close your emulator entirely, restart it and re-run these steps.
    2. If it says `Must use a version of bizhawk 2.3.1 or higher`, double-check your Bizhawk version by clicking **
       Help** -> **About**.
```

File: worlds/tloz/docs/multiworld_en.md [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+
+## Play the game
+
+When the client shows both NES and server are connected you are good to go. You can check the connection status of the
+NES at any time by running `/nes`
+
+### Other Client Commands
+
+All other commands may be found on the [Archipelago Server and Client Commands Guide](/tutorial/Archipelago/commands/en)
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:91]:
```suggestion
All other commands may be found on the [Archipelago Server and Client Commands Guide.](/tutorial/Archipelago/commands/en)
```

File: worlds/tloz/docs/multiworld_en.md [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+4. Navigate to the location you installed Archipelago to. Open data/lua/TLOZ/tloz_connector.lua
+    1. If it gives a `NLua.Exceptions.LuaScriptException: .\socket.lua:13: module 'socket.core' not found:` exception
+       close your emulator entirely, restart it and re-run these steps
+    2. If it says `Must use a version of bizhawk 2.3.1 or higher`, double-check your Bizhawk version by clicking **
+       Help** -> **About**
+
+## Play the game
+
+When the client shows both NES and server are connected you are good to go. You can check the connection status of the
+NES at any time by running `/nes`
```
Comment (SoldierofOrder) [inline worlds/tloz/docs/multiworld_en.md:87]:
```suggestion
When the client shows both NES and server are connected, you are good to go. You can check the connection status of the
NES at any time by running `/nes`.
```

**Review by SoldierofOrder — APPROVED** [review]
I only examined the user-facing docs files (setup guide and info page). Mostly, I just saw punctuation stuff; looks good to go otherwise.

File: worlds/tloz/__init__.py [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+                    elif item_class == ItemClassification.trap:
+                        item_price = item_price * 2
+                rom_data[price_location] = item_price
+            if location.name == "Take Any Item Right":
+                # Same story as above: bit 6 is what makes this a Take Any cave
+                item_id = item_id | 0b01000000
+            rom_data[location_id] = item_id
+        
+        # We shuffle the tiers of rupee caves. Caves that shared a value before still will.
+        secret_caves = self.multiworld.random.sample(sorted(secret_money_ids), 3)
```
Comment (Alchav) [inline worlds/tloz/__init__.py:229]: Use of multiworld.random in generate_output.

multiworld.random is now made unusable before generate_output so this crashes. local random object should be used

File: worlds/tloz/Rules.py [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+
+    for location in tloz_world.levels[9].locations:
+        add_rule(world.get_location(location.name, player),
+                 lambda state: state.has("Triforce Fragment", player, 8) and
+                               state.has_group("swords", player))
+
+    # Yes we are looping this range again for Triforce locations. No I can't add it to the boss event loop
+    for level in range(1, 9):
+        add_rule(world.get_location(f"Level {level} Triforce", player),
+                 lambda state: state.has(f"Boss {level} Defeated", player))
```
Comment (Alchav) [inline worlds/tloz/Rules.py:92]: Creating a lambda in a loop like this causes all these lambdas to end up referenced with `level` as it is at the end of the loop, which causes them all to logically require Boss 9 Defeated. This would be a way to do this correctly:
```
        add_rule(world.get_location(f"Level {level} Triforce", player),
                 lambda state, l=level: state.has(f"Boss {l} Defeated", player))
```

File: worlds/tloz/Rules.py [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+    player = tloz_world.player
+    world = tloz_world.multiworld
+
+    # Boss events for a nicer spoiler log play through
+    for level in range(1, 9):
+        boss = world.get_location(f"Level {level} Boss", player)
+        boss_event = world.get_location(f"Level {level} Boss Status", player)
+        status = tloz_world.create_event(f"Boss {level} Defeated")
+        boss_event.place_locked_item(status)
+        add_rule(boss_event, lambda state: state.can_reach(boss, "Location", player))
```
Comment (Alchav) [inline worlds/tloz/Rules.py:20]: as below, this is creating the same rule for all locations. Something like this should fix it:
`add_rule(boss_event, lambda state, b=boss: state.can_reach(b, "Location", player))`

File: worlds/tloz/Rules.py [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+
+    # No dungeons without weapons except for the dangerous weapon locations if we're dangerous, no unsafe dungeons
+    for i, level in enumerate(tloz_world.levels[1:10]):
+        for location in level.locations:
+            if world.StartingPosition[player] < 1 or location.name not in dangerous_weapon_locations:
+                add_rule(world.get_location(location.name, player),
+                         lambda state: state.has_group("weapons", player))
+            if i > 0:  # Don't need an extra heart for Level 1
+                add_rule(world.get_location(location.name, player),
+                         lambda state: state.has("Heart Container", player, i) or
```
Comment (Alchav) [inline worlds/tloz/Rules.py:30]: yet again using i from a `for` loop unsafely, going to be 9 or whatever for all of them

**Comment by Alchav** [comment]
Went ahead and sent a PR to this branch to fix the lambda issues

**Comment by Rosalie-A** [comment]
> Went ahead and sent a PR to this branch to fix the lambda issues

oops I didn't see this until after I'd already written and pushed your suggested changes. Don't take my missing it as anything other than my being absent-minded

**Comment by Alchav** [comment]
> 

It's fine as long as it's fixed :)

File: worlds/tloz/Locations.py [hunk 30]
```diff
@@ -0,0 +1,350 @@
...
+        "Level 9 Bomb Drop (Blue Lanmolas)", "Level 9 Bomb Drop (Gels Lake)",
+        "Level 9 Bomb Drop (Like Likes and Zols Corridor)", "Level 9 Bomb Drop (Patra Northeast)",
+        "Level 9 Bomb Drop (Vires)", "Level 9 Rupee Drop (Wizzrobes West Island)",
+        "Level 9 Rupee Drop (Red Lanmolas)", "Level 9 Rupee Drop (Keese Southwest)",
+        "Level 9 Rupee Drop (Keese Central Island))", "Level 9 Rupee Drop (Wizzrobes Central)",
+        "Level 9 Rupee Drop (Wizzrobes North Island)", "Level 9 Rupee Drop (Gels East)"
+    ]
+]
+
+all_level_locations = []
```
Comment (alwaysintreble) [inline worlds/tloz/Locations.py:90]: `all_level_locations = [location for level in level_locations for location in level]

File: worlds/tloz/Locations.py [hunk 31]
```diff
@@ -0,0 +1,350 @@
...
+    "Level 6 Key Drop (Wizzrobes North Island)": 0x29,
+    "Level 6 Key Drop (Wizzrobes North Stream)": 0x1A,
+    "Level 6 Key Drop (Vires)": 0x2D,
+    "Level 6 Bomb Drop (Wizzrobes)": 0x3C,
+    "Level 6 Rupee Drop (Wizzrobes)": 0x28
+}
+
+floor_location_game_ids_early = {}
+floor_location_game_ids_late = {}
+for key, value in floor_location_game_offsets_early.items():
```
Comment (alwaysintreble) [inline worlds/tloz/Locations.py:194]: `floor_location_game_ids_early = {key, value + Rom.first_quest_dungeon_items_early for key, value in floor_location_game_offsets_early.items()}

File: worlds/tloz/Locations.py [hunk 32]
```diff
@@ -0,0 +1,350 @@
...
+    "Level 9 Rupee Drop (Keese Central Island))": 0x34,
+    "Level 9 Rupee Drop (Wizzrobes Central)": 0x44,
+    "Level 9 Rupee Drop (Wizzrobes North Island)": 0x15,
+    "Level 9 Rupee Drop (Gels East)": 0x26
+}
+
+for key, value in floor_location_game_offsets_late.items():
+    floor_location_game_ids_late[key] = value + Rom.first_quest_dungeon_items_late
+
+dungeon_items = {**floor_location_game_ids_early, **floor_location_game_ids_late}
```
Comment (alwaysintreble) [inline worlds/tloz/Locations.py:260]: do you use all 3 of these dictionaries separately? seems like it'd be better to just build a single dictionary with a single generator using `collections.ChainMap`

File: worlds/tloz/Locations.py [hunk 33]
```diff
@@ -0,0 +1,350 @@
...
+overworld_locations = [
+    "Starting Sword Cave",
+    "White Sword Pond",
+    "Magical Sword Grave",
+    "Letter Cave",
+    "Armos Knights",
+    "Ocean Heart Container"
+]
+
+underworld1_locations = [*floor_location_game_offsets_early.keys()]
```
Comment (alwaysintreble) [inline worlds/tloz/Locations.py:337]: `list(floor_location_game_offsets_early)`

File: worlds/tloz/Rules.py [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+
+if TYPE_CHECKING:
+    from . import TLoZWorld
+
+def set_rules(tloz_world: "TLoZWorld"):
+    player = tloz_world.player
+    world = tloz_world.multiworld
+
+    # Boss events for a nicer spoiler log play through
+    for level in range(1, 9):
```
Comment (alwaysintreble) [inline worlds/tloz/Rules.py:15]: python ranges are end exclusive so this only does 1-8 if that matters here

File: worlds/tloz/Rules.py [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+
+    stepladder_locations = [
+        "Ocean Heart Container", "Level 4 Triforce", "Level 4 Boss",
+        "Level 4 Map", "Level 4 Key Drop (Keese North)"
+    ]
+    for location in stepladder_locations:
+        add_rule(world.get_location("Ocean Heart Container", player),
+                 lambda state: state.has("Stepladder", player))
+
+    if world.StartingPosition[player] != 2:
```
Comment (alwaysintreble) [inline worlds/tloz/Rules.py:108]: I'd recommend importing the actual Option class and using the value, or using a string comparison, instead of a hardcoded number comparison that doesn't mean anything when you reread this code in 3 months.

File: worlds/tloz/Rules.py [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+        add_rule(world.get_location(location.name, player),
+                 lambda state: state.has("Raft", player) or state.has("Recorder", player))
+    for location in tloz_world.levels[7].locations:
+        add_rule(world.get_location(location.name, player),
+                 lambda state: state.has("Recorder", player))
+    for location in tloz_world.levels[8].locations:
+        add_rule(world.get_location(location.name, player),
+                 lambda state: state.has("Bow", player))
+
+    add_rule(world.get_location("Potion Shop Item Left", player),
```
Comment (alwaysintreble) [inline worlds/tloz/Rules.py:130]: Could these be a Region?

File: worlds/tloz/__init__.py [hunk 37]
```diff
@@ -0,0 +1,308 @@
...
+        ganon = self.multiworld.get_location("Ganon", self.player)
+        ganon.place_locked_item(self.create_event("Triforce of Power"))
+        add_rule(ganon, lambda state: state.has("Silver Arrow", self.player) and state.has("Bow", self.player))
+
+        self.multiworld.get_location("Zelda", self.player).place_locked_item(self.create_event("Rescued Zelda!"))
+        add_rule(self.multiworld.get_location("Zelda", self.player),
+                 lambda state: ganon in state.locations_checked)
+
+        self.multiworld.completion_condition[self.player] = lambda state: state.has("Rescued Zelda!", self.player)
+        generate_itempool(self)
```
Comment (alwaysintreble) [inline worlds/tloz/__init__.py:148]: create items in the function dedicated to it please.

**Review by alwaysintreble — COMMENTED** [review]
mostly lgtm

File: worlds/tloz/__init__.py (hunk 37 above)
Comment (t3hf1gm3nt) [reply worlds/tloz/__init__.py:148]: since all of the item creation is done in this function that is in a whole other file (ItemPool.py), sticking it in create_items seemed redundant to me? unless theres something fundamental im not understanding

File: worlds/tloz/Rules.py [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+             lambda state: state.has("Heart Container", player, 2))
+    add_rule(world.get_location("Magical Sword Grave", player),
+             lambda state: state.has("Heart Container", player, 9))
+
+    stepladder_locations = [
+        "Ocean Heart Container", "Level 4 Triforce", "Level 4 Boss",
+        "Level 4 Map", "Level 4 Key Drop (Keese North)"
+    ]
+    for location in stepladder_locations:
+        add_rule(world.get_location("Ocean Heart Container", player),
```
Comment (Alchav) [inline worlds/tloz/Rules.py:105]: I'm guessing this should be 
`add_rule(world.get_location(location, player),`

File: worlds/tloz/__init__.py (hunk 37 above)
Comment (Berserker66) [reply worlds/tloz/__init__.py:149]: You're doing it later than expected, which has lead to bugs before and in the future AP is likely to reject your world for it.

**Review by Berserker66 — APPROVED** [review]

**Comment by Berserker66** [comment]
Merging this for now, but I'd like to see the code quality concerns that were commented addressed soon. Almost inevitably someone will copy from your world.
